pkg/kubelet/kubelet_node_status.go

```diff
@@ -755,6 +765,33 @@ func (kl *Kubelet) defaultNodeStatusFuncs() []func(context.Context, *v1.Node) er
 	return setters
 }
 
+// validateWritableDirectory is used to ensure that we can write to the directories that we expect to.  During runtime
+// filesystems can be re-mounted as read-only or experience other conditions that prevent writing to them.  This readiness
+// check causes the node to go NotReady in those situations.
+func validateWritableDirectory(label, dirName string) func() error {
+	return func() error {
+		info, err := fs.Info(dirName)
+		if err != nil {
+			return fmt.Errorf("unable to determine filesystem info for %s directory %s, %w", label, dirName, err)
+		}
+		if info.ReadOnly {
+			return fmt.Errorf("%s directory %s is mounted read-only", label, dirName)
+		}
+
+		// Actually writing a file can detect a rare condition with xfs where you have a r/w mounted filesystem,
+		// plenty of free space and inodes reported, but that free space is fragmented enough that there isn't enough
+		// contiguous free space for new inodes to be dynamically allocated. Anything that requires a new inode, such as
+		// creation of a new file returns ENOSPC in that case.
+		f, err := os.CreateTemp(dirName, ".validate-writable-*")
```

Comment:
How frequency do we validate the WritableDirectory ? Trying to understand the how much the iops will this validate function take. 

Looking at the node problem detector, it checking the condition by scanning the dmesg log: 
https://github.com/kubernetes/node-problem-detector/blob/master/config/kernel-monitor.json#L76

Reply:
It runs every 20 seconds.  NPD looks for this, but it may not be a relevant filesystem and it doesn't detect other odd conditions which cause a filesystem to be unwritable even though it's not readonly.